### PR TITLE
EarningAchievementEvent

### PR DESCRIPTION
**Pull request by Ruemena**
might need a bit of testing

### CODE REVIEW

**Comment by louis1706** [comment]
image
i would just like to remind than rule exist

**Comment by Ruemena** [comment]
this isn't granting achievements, this can only deny achievements

**Comment by louis1706** [comment]
> this isn't granting achievements, this can only deny achievements

yup it's was what i was veryfying

**Comment by Thundermaker300** [comment]
I'm cool with an event when an achievement is granted. Not sure we should be denying granting achievements though

**Comment by Ruemena** [comment]
i mean there's valid uses for it, maybe you have a custom item and don't want it to grant the achievement (i.e. micro) or want to not give it during events

**Comment by Thundermaker300** [comment]
That's fair actually. Resolve conflicts and test and let's get it merged 👍

**Comment by Ruemena** [comment]
resolved conflicts, need the tests to be run tho

**Review by Thundermaker300 — APPROVED** [review]
Transpiler still needs reviewed

**Review by louis1706 — APPROVED** [review]

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]
Fix builds errors

File: Exiled.Events/Patches/Events/Player/EarningAchievement.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    using Mirror;
+
+    using static HarmonyLib.AccessTools;
+
+    /// <summary>
+    ///     Patch the <see cref="AchievementHandlerBase.ServerAchieve"/>.
+    ///     Adds the <see cref="Handlers.Player.EarningAchievement"/> event.
+    /// </summary>
+    [HarmonyPatch(typeof(AchievementHandlerBase), nameof(AchievementHandlerBase.ServerAchieve))]
```
Comment (louis1706) [inline Exiled.Events/Patches/Events/Player/EarningAchievement.cs:29]: Missing the new EventPatch argument (look other event)

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Events/Player/EarningAchievement.cs (hunk 1 above)
Comment (Ruemena) [reply Exiled.Events/Patches/Events/Player/EarningAchievement.cs:30]: done

**Comment by Ruemena** [comment]
need tests to be run but i fixed the build errors

**Comment by Ruemena** [comment]
nvm i did not

**Comment by Thundermaker300** [comment]
im so proud :heart:

**Comment by Ruemena** [comment]
next update is gonna break the tests and im gonna have to have them run 5 times again

> im so proud ❤️

File: Exiled.Events/Patches/Events/Player/EarningAchievement.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+                    // Handlers.Player.OnEarningAchievement(ev);
+                    new(OpCodes.Call, Method(typeof(Handlers.Player), nameof(Handlers.Player.OnEarningAchievement))),
+
+                    // if (!ev.Isallowed) return;
+                    new(OpCodes.Callvirt, PropertyGetter(typeof(EarningAchievementEventArgs), nameof(EarningAchievementEventArgs.IsAllowed))),
+                    new(OpCodes.Brfalse, ret),
+                });
+
+            newInstructions[newInstructions.Count - 1].WithLabels(ret);
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/EarningAchievement.cs:72]: `newInstructions[newInstructions.Count - 1].labels.Add(ret);`

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

**Review by louis1706 — APPROVED** [review]

**Pull request merged**